### PR TITLE
[Feature 53] Basic Setup of the GameBoard

### LINKED ISSUE #53: [51 ]Enhance Gameboard Interface - Interface Adapter Level
I will be working on enhancing the gameboard interface to improve the user experience and introduce new features for our Monopoly game. The goal is to follow the principles of Clean Architecture to ensure maintainability and flexibility in our codebase.

**Tasks:**

- [ ]  **User Info Blocks:**
   - Add user info blocks to the left-side bar of the gameboard. These blocks will display relevant information about each player, such as their name, avatar, and current game statistics.

- [x]  **Divide Game Board into Grids:**
   - Modify the middle section of the gameboard to be divided into grids, reflecting the layout of a Monopoly board. Each grid will represent a specific property or space on the board.

- [x]  **Announcement Thread:**
   - Create an announcement thread on the right-side bar of the gameboard. This thread will display important announcements, game events, and notifications for all players to see.

**Clean Architecture - Interface Adapter Level:**
The tasks mentioned above fall under the Interface Adapter Level of Clean Architecture. This layer is responsible for converting data and events from the core business logic into formats that the user interface can understand, and vice versa. It ensures that the gameboard's presentation and interaction with users are handled separately from the core Monopoly game logic. By implementing these enhancements at the Interface Adapter Level, we can make our gameboard more flexible, easy to maintain, and adaptable to future changes without affecting the core game mechanics.

### PR DESCRIPTION
**Pull request by Noahkong12**
### **Architecture Breakdown**
**Entry Point:**

The main method serves as the entry point and instantiates the game board inside the Event Dispatch Thread (ensuring thread safety for Swing components).
**Main Game Frame: GameBoard**

Inherits from JFrame.
Sets basic window properties like title, size, default close operation, etc.
Initializes the main game board and sidebars.
Uses BorderLayout to position the game board centrally with sidebars on its left and right.

**Game Board Panel: GameBoardPanel**
Responsible for rendering the main game board, including the title, board layout, properties, destiny cards, etc.
Overrides the paintComponent method to customize the painting behavior.
Contains helper methods (drawGameTitle, drawBoard, drawProperties, etc.) to decompose the drawing tasks, ensuring the Single Responsibility Principle (SRP).
Uses an array, propertyColors, to manage property colors, with encapsulation in place to modify colors via setColorForProperty.

**Sidebars: LeftSidebar & RightSidebar**
Both inherit from JPanel.
Use fixed width for consistent layout.
Override the paintComponent method to render sidebars.
LeftSidebar focuses on displaying player details. It renders a blue background and a title.
RightSidebar is magenta-colored and divides its space for game threads and other yet-to-be-defined features.
### **Design Rule and Principles Applied**
**Encapsulation:** Through the use of private inner classes and methods, the internal structure and methods of the game board are hidden, providing a cleaner external interface.

**Single Responsibility Principle (SRP):** 
The code splits different responsibilities into different methods (like drawGameTitle, drawBoard, etc.) and classes (LeftSidebar, RightSidebar, etc.).

**Open/Closed Principle:** 
Classes like GameBoardPanel are open for extension (e.g., more drawing methods can be added) but closed for modification, especially with the use of final for some constants.

**DRY (Don't Repeat Yourself):** 
Repeated actions like drawing properties and destiny card spaces are modularized into separate methods to avoid repetition.

**Separation of Concerns:** 
Each class and method handles a specific part of the game UI. The game board focuses on the game's main drawing, the sidebars on auxiliary info, and so forth.

**Dependency Inversion:** 
While this principle is more applicable to bigger systems where high-level modules shouldn't depend on low-level ones, the structure here gives an inclination towards it. For instance, higher-level components like the GameBoard do not depend directly on lower-level drawing details but on abstracted methods.

**Polymorphism:** 
Through the use of inheritance (e.g., extending JPanel), classes can override parent methods (paintComponent) to provide custom behaviors.

### **Finally, feel free to comment if you have questions regarding the code**

### CODE REVIEW

**Comment by snow-QingYang** [comment]
I think we should delete GameMain. It seems that it is useless now.

**Comment by AlwynYin** [comment]
Can you rebase this branch onto main? Some files seem to be outdated.

**Comment by snow-QingYang** [comment]
Also, It seems that the blocks in the corner are hidden by other UI. Can you fix it ? or Just write EC instead of Exam Center

**Comment by snow-QingYang** [comment]
> Can you rebase this branch onto main? Some files seem to be outdated.

yea, I agree with Alwyn

**Comment by snow-QingYang** [comment]
> Also, It seems that the blocks in the corner are hidden by other UI. Can you fix it ? or Just write EC instead of Exam Center

and TTC instead of TTC station

**Comment by Noahkong12** [comment]
No problem, I will fix the text issues in the next version of the board. Regarding the GameMain, yea, it is useless since I already put a "main" in my GameBoard JAVAclass temporily to see how the board looks like.

**Comment by snow-QingYang** [comment]
> No problem, I will fix the text issues in the next version of the board. Regarding the GameMain, yea, it is useless since I already put a "main" in my GameBoard JAVAclass temporarily to see what the board looks like.

That's great! I hope to see the new version of our monopoly! I think next task is to add keyboard and mouse input into UI

**Comment by snow-QingYang** [comment]
Also, I think we can add some photos for exam center or other buildings.

**Comment by Noahkong12** [comment]
I’ve thought about adding pictures actually, but seems that the grid is too small for a sufficiently representative and clear picture of those locations. But I will try, thanks for the advice!

**Comment by AlwynYin** [comment]
> I’ve thought about adding pictures actually, but seems that the grid is too small for a sufficiently representative and clear picture of those locations. But I will try, thanks for the advice!

maybe we can display at the sidebar the picture of the block that current player is on?

**Comment by snow-QingYang** [comment]
I have found a new problem. I found that the game window has a disproportionate scaling after enlarging. Can you fix that?

**Comment by snow-QingYang** [comment]
> 

maybe you can add absolute position to each UI

File: src/main/java/UI/GameBoard.java [hunk 1]
```diff
@@ -35,17 +35,144 @@ private class GameBoardPanel extends JPanel {
...
+            FontMetrics fm = g.getFontMetrics();
+            int titleWidth = fm.stringWidth("Monopoly: Adventure at UofT");
+            int titleHeight = fm.getAscent();  // This gives the height from base to top of the font
+
+            // Calculate the position to center the title
+            int x = (getWidth() - titleWidth) / 2;
+            int y = titleHeight + 20;  // A little padding from the top
+
+            // Set color for the title
+            g.setColor(new Color(173, 216, 230));  // Light blue color
```
Comment (NeverSoGoodStart) [inline src/main/java/UI/GameBoard.java:60]: It's better to put specific colors that you plan to use on the top of the file, so it allows for easier change later on.

File: src/main/java/UI/GameBoard.java [hunk 2]
```diff
@@ -62,16 +197,32 @@ public Dimension getPreferredSize() {
...
+        private static final int THREAD_SECTION_HEIGHT = 900 / 2;  // bottom half of the sidebar
+        private static final int THREAD_TITLE_PADDING = 20;  // distance from the top of the thread section
 
         @Override
         protected void paintComponent(Graphics g) {
             super.paintComponent(g);
 
             // Draw the right sidebar
-            g.setColor(Color.RED);
+            g.setColor(Color.MAGENTA);
```
Comment (NeverSoGoodStart) [inline src/main/java/UI/GameBoard.java:211]: Same color problem with this

**Review by NeverSoGoodStart — COMMENTED** [review]
It mostly looks good to me so far, just some of the colors I think can be written as global variables and declared on top of the file. 

If you think the current window size is good, and you're running out of time, you can also just add: 
`setResizable(false)` in `GameBoard`'s constructor

**(The following is not really urgent)**
However, one small problem I can tell is: 
All of the numbers of the UI are like, magical numbers that you have tried out and worked on your screen. 

If I run the program on a 4k monitor, it might just look very weird (I'm guessing tho)

What I suggest doing is, probably get the screen display size by using `java.awt.getScreenSize()`, then use some mathematics to figure out how big the window has be. 

Something like: If the game window is in a square, then I take the minimum of screen width / height. 

Then I take `window_width = 0.6 * min(screen_width, screen_height)`, and take `math.ceil()` or `math.floor()` to ensure that the display is not too big / too small on all different types of display. In other words, make it adaptive. 

Then, once you have figured out how big the window will be, you can set almost everything based on that. 

For example, if you think in 600 * 600, font size 30 is good, you can use that proportion. 
E.g., 30 / 600 = 0.05, so we can set the font size to be `0.05 * window_width`, where `window_width` is calculated in the previous step. 

Similar with most of the other things.

**Review by NeverSoGoodStart — APPROVED** [review]
🆒

**Review by AlwynYin — APPROVED** [review]

**Review by hecateli — APPROVED** [review]

**Pull request merged**